core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java

```diff
@@ -0,0 +1,438 @@
+              keysToExclude, includeLoader, operation, callback, StreamRequestCommand.Type.TERMINAL,
+              earlyTerminatePredicate);
+   }
+
+   @Override
+   public <R> UUID remoteStreamOperationRehashAware(boolean parallelDistribution, boolean parallelStream,
+           ConsistentHash ch, Set<Integer> segments, Set<K> keysToInclude, Map<Integer, Set<K>> keysToExclude,
+           boolean includeLoader, TerminalOperation<R> operation, ResultsCallback<R> callback,
+           Predicate<? super R> earlyTerminatePredicate) {
+      return remoteStreamIgnoreKeyOperation(parallelDistribution, parallelStream, ch, segments, keysToInclude,
+              keysToExclude, includeLoader, operation, callback, StreamRequestCommand.Type.TERMINAL_REHASH,
+              earlyTerminatePredicate);
+   }
+
+   private <R> UUID remoteStreamIgnoreKeyOperation(boolean parallelDistribution, boolean parallelStream,
+           ConsistentHash ch, Set<Integer> segments, Set<K> keysToInclude, Map<Integer, Set<K>> keysToExclude,
+           boolean includeLoader, TerminalOperation<R> operation, ResultsCallback<R> callback,
+           StreamRequestCommand.Type type, Predicate<? super R> earlyTerminatePredicate) {
+      Map<Address, Set<Integer>> targets = determineTargets(ch, segments);
+      UUID uuid = UUID.randomUUID();
+      if (!targets.isEmpty()) {
+         log.tracef("Performing remote operations %s for id %s", targets, uuid);
+         RequestTracker<R> tracker = new RequestTracker<>(callback, targets, earlyTerminatePredicate);
+         currentlyRunning.put(uuid, tracker);
+         if (parallelDistribution) {
+            submitAsyncTasks(uuid, targets, keysToExclude, parallelStream, keysToInclude, includeLoader, type,
+                    operation);
+         } else {
+            for (Map.Entry<Address, Set<Integer>> targetInfo : targets.entrySet()) {
+               // TODO: what if this throws exception?
+               Set<Integer> targetSegments = targetInfo.getValue();
+               Set<K> keysExcluded = determineExcludedKeys(keysToExclude, targetSegments);
+               rpc.invokeRemotely(Collections.singleton(targetInfo.getKey()),
+                       factory.buildStreamRequestCommand(uuid, parallelStream, type, targetSegments, keysToInclude,
+                               keysExcluded, includeLoader, operation), rpc.getDefaultRpcOptions(true));
+            }
+         }
+      } else {
+         log.tracef("Not performing any remote operations for id %s as no valid targets found", uuid);
+      }
+      return uuid;
+   }
+
+   @Override
+   public <R> UUID remoteStreamOperation(boolean parallelDistribution, boolean parallelStream, ConsistentHash ch,
+           Set<Integer> segments, Set<K> keysToInclude, Map<Integer, Set<K>> keysToExclude, boolean includeLoader,
+           KeyTrackingTerminalOperation<K, R, ?> operation, ResultsCallback<Collection<R>> callback) {
```

Comment:
I think `remoteStreamOperation` and `remoteStreamIgnoreKeyOperation` could be unified.